<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [19]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
!ls -lh

--2019-11-04 18:51:44--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2019-11-04 18:51:45 (126 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]

total 48K
-rw-r--r-- 1 root root  18K May 30  1989 house-votes-84.data
-rw-r--r-- 1 root root  18K May 30  1989 house-votes-84.data.1
-rw-r--r-- 1 root root  653 Nov  4 18:36 index.html
drwxr-xr-x 1 root root 4.0K Oct 25 16:58 sample_data


In [33]:
import pandas as pd
import numpy as np
import seaborn as sns

column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df = pd.read_csv("house-votes-84.data", header=None, names=column_headers)
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
rep = df[df["party"] == "republican"]
dem = df[df["party"] == "democrat"]

In [0]:
rep = rep.drop(columns="party")
dem = dem.drop(columns="party")

In [0]:
rep = rep.replace({"n": 0, "y": 1, "?": np.NaN})

In [0]:

dem = dem.replace({"n": 0, "y": 1, "?": np.NaN})

In [38]:
rep.head()

handicapped-infants  water-project  budget  ...  crime  duty-free  south-africa
0                   0.0            1.0     0.0  ...    1.0        0.0           1.0
1                   0.0            1.0     0.0  ...    1.0        0.0           NaN
7                   0.0            1.0     0.0  ...    1.0        NaN           1.0
8                   0.0            1.0     0.0  ...    1.0        0.0           1.0
10                  0.0            1.0     0.0  ...    1.0        0.0           0.0

[5 rows x 16 columns]

In [39]:
dem.head()

handicapped-infants  water-project  budget  ...  crime  duty-free  south-africa
2                  NaN            1.0     1.0  ...    1.0        0.0           0.0
3                  0.0            1.0     1.0  ...    0.0        0.0           1.0
4                  1.0            1.0     1.0  ...    1.0        1.0           1.0
5                  0.0            1.0     1.0  ...    1.0        1.0           1.0
6                  0.0            1.0     0.0  ...    1.0        1.0           1.0

[5 rows x 16 columns]

In [40]:
rep.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [41]:
dem.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [0]:
from scipy.stats import ttest_ind

class ColumnCompare:
  
  def __init__(self, x1, x2):
    """
      @type x1: DataFrame
      @type x2: DataFrame
    """
    self.__x1 = x1
    self.__x2 = x2

  def test(self, column_name):
    """
      @type column_name: str
      @return: Ttest_indResult of the result of the test for the given column.
    """
    return ttest_ind(self.__x1[column_name], self.__x2[column_name], nan_policy="omit")

In [0]:
comp = ColumnCompare(dem, rep)

### Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

1) $\bar{x}_1 == \bar{x}_2$

2) $\bar{x}_1 \neq \bar{x}_2$

3) 99% Confidence level

In [56]:
comp.test("aid-to-contras")

Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54)

4) T-statistic: 18.05

5) P-value: ~0

Due to P < 0.01 we **reject** the null hypothesis. Support for the bill is not equal.

### Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

1) Null hypothesis: There is no difference in support between democrats and republicans on "education."

$\bar{x}_1 == \bar{x}_2$

2) Alternate hypothesis: There is a difference in support between democrats and republicans on "education".

$\bar{x}_1 \neq \bar{x}_2$

3) 99% confidence level

In [57]:
comp.test("education")

Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64)

4) T-statistic: 20.5

5) P-value: ~0

Because P < 0.01 we **reject** the null hypothesis. Support for the bill is not equal.

### Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

1) Null hypothesis: 

$\bar{x}_{1} == \bar{x}_{2}$

2) Alternative Hypothesis: 

$\bar{x}_{1} \neq \bar{x}_{2}$

Levels of support between the two parties will differ. 

3) 90% Confidence Level

In [58]:
comp.test("immigration")

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

4) T-statistic: 1.74

5) P value: 0.083

Based on a P value < 0.1, we **reject** the null hypothesis. Support for the bills is different.

### Next try south africa 

In [59]:
comp.test("south-africa")

Ttest_indResult(statistic=6.849454815841208, pvalue=3.652674361672226e-11)

Based on P < 0.1, we **reject** the null hypothesis. Support for the bills is different.

### Next try water project

In [60]:
comp.test("water-project")

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

Based on P > 0.1, we **fail to reject** the null hypothesis. Support for the bills is similar.

In [79]:
df = pd.read_csv("ufc_submission_stats.csv", index_col="year")
df.head()

Rear Naked Choke  Armbar  Guillotine  Triangle Choke  Arm Triangle
year                                                                    
2010                11       6           9               6             4
2011                22       3          14               4             2
2012                30      12          14               4             2
2013                17      10          15               6             8
2014                41       8          20               3             6

In [80]:
df.describe()


Rear Naked Choke     Armbar  Guillotine  Triangle Choke  Arm Triangle
count          9.000000   9.000000    9.000000        9.000000      9.000000
mean          31.333333   8.444444   14.888889        4.444444      5.222222
std           12.051971   2.743680    4.702245        1.589899      2.279132
min           11.000000   3.000000    9.000000        2.000000      2.000000
25%           22.000000   7.000000   11.000000        3.000000      4.000000
50%           36.000000  10.000000   14.000000        4.000000      5.000000
75%           41.000000  10.000000   16.000000        6.000000      7.000000
max           45.000000  12.000000   24.000000        6.000000      8.000000

In [0]:
df = df.div(df.sum(axis=1), axis=0)

In [82]:
df.describe()

Rear Naked Choke    Armbar  Guillotine  Triangle Choke  Arm Triangle
count          9.000000  9.000000    9.000000        9.000000      9.000000
mean           0.472293  0.134016    0.235038        0.076318      0.082335
std            0.100618  0.043302    0.052118        0.042459      0.034657
min            0.303571  0.066667    0.150685        0.031250      0.032258
25%            0.476744  0.102564    0.202532        0.038462      0.058140
50%            0.488889  0.136986    0.250000        0.069767      0.078125
75%            0.534247  0.166667    0.267857        0.088889      0.101266
max            0.569620  0.193548    0.311111        0.166667      0.142857

In [0]:
from scipy.stats import ttest_1samp

In [83]:
ttest_1samp(df["Rear Naked Choke"], 0.47)

Ttest_1sampResult(statistic=0.06837375023198605, pvalue=0.947166183232409)

Due to a high p-value we **fail to reject** the hypothesis that the median is 0.47. 

In [84]:
ttest_1samp(df["Armbar"], 0.13)

Ttest_1sampResult(statistic=0.27820560267067224, pvalue=0.7879097183592841)

Due to a high p-value we **fail to reject** the hypothesis that the median is 0.13.

In [85]:
ttest_1samp(df["Triangle Choke"], 0.076)

Ttest_1sampResult(statistic=0.02245285747266912, pvalue=0.9826366452424405)

Due to a high p-value we **fail to reject** the hypothesis that the median is 0.076.